### PR TITLE
[SPARK-7824][SQL]Collapsing operator reordering and constant folding into a single batch  to push down the single side.

### PR DESCRIPTION
**Pull request by pzzs**
SQL

```
select * from tableA join tableB on (a > 3 and b = d) or (a > 3 and b = e)
```

Plan before modify

```
== Optimized Logical Plan ==
Project [a#293,b#294,c#295,d#296,e#297]
 Join Inner, Some(((a#293 > 3) && ((b#294 = d#296) || (b#294 = e#297))))
  MetastoreRelation default, tablea, None
  MetastoreRelation default, tableb, None
```

Plan after modify

```
== Optimized Logical Plan ==
Project [a#293,b#294,c#295,d#296,e#297]
 Join Inner, Some(((b#294 = d#296) || (b#294 = e#297)))
  Filter (a#293 > 3)
   MetastoreRelation default, tablea, None
  MetastoreRelation default, tableb, None
```

CombineLimits ==> Limit(If(LessThan(ne, le), ne, le), grandChild) and LessThan is in BooleanSimplification ,  so CombineLimits  must before BooleanSimplification and BooleanSimplification must before PushPredicateThroughJoin.

### CODE REVIEW

**Comment by DoingDone9** [comment]
This optimizer can void CartesianProduct
Tables
```
tableA        tableB        tableC
a   int       c    int      f    int    
b   int       d    int      g    int
              e    int
```

SQL
```
select * from tableA, tableB, tableC where (d = f and  a = c and b = d) or (d = f and a = c and b = e)
```

Plan before modify
```
== Optimized Logical Plan ==
Project [a#306,b#307,c#308,d#309,e#310,f#311,g#312]
 Join Inner, Some((((d#309 = f#311) && (a#306 = c#308)) && ((b#307 = d#309) || (b#307 = e#310))))
  Join Inner, None
   MetastoreRelation default, tablea, None
   MetastoreRelation default, tableb, None
  MetastoreRelation default, tablec, None

== Physical Plan ==
Project [a#306,b#307,c#308,d#309,e#310,f#311,g#312]
 Filter ((a#306 = c#308) && ((b#307 = d#309) || (b#307 = e#310)))
  ShuffledHashJoin [d#309], [f#311], BuildRight
   Exchange (HashPartitioning 5), []
    CartesianProduct
     HiveTableScan [a#306,b#307], (MetastoreRelation default, tablea, None), None
     HiveTableScan [c#308,d#309,e#310], (MetastoreRelation default, tableb, None), None
   Exchange (HashPartitioning 5), []
    HiveTableScan [f#311,g#312], (MetastoreRelation default, tablec, None), None
```

Plan after modify
```
== Optimized Logical Plan ==
Project [a#306,b#307,c#308,d#309,e#310,f#311,g#312]
 Join Inner, Some((d#309 = f#311))
  Join Inner, Some(((a#306 = c#308) && ((b#307 = d#309) || (b#307 = e#310))))
   MetastoreRelation default, tablea, None
   MetastoreRelation default, tableb, None
  MetastoreRelation default, tablec, None

== Physical Plan ==
Project [a#306,b#307,c#308,d#309,e#310,f#311,g#312]
 ShuffledHashJoin [d#309], [f#311], BuildRight
  Exchange (HashPartitioning 5), []
   Filter ((b#307 = d#309) || (b#307 = e#310))
    ShuffledHashJoin [a#306], [c#308], BuildRight
     Exchange (HashPartitioning 5), []
      HiveTableScan [a#306,b#307], (MetastoreRelation default, tablea, None), None
     Exchange (HashPartitioning 5), []
      HiveTableScan [c#308,d#309,e#310], (MetastoreRelation default, tableb, None), None
  Exchange (HashPartitioning 5), []
   HiveTableScan [f#311,g#312], (MetastoreRelation default, tablec, None), None
```

**Comment by DoingDone9** [comment]
@marmbrus @scwf

**Comment by scwf** [comment]
Get it. The root cause here is we moved `Filter Pushdown` before `ConstantFolding` in https://github.com/DoingDone9/spark/commit/4459514497eb76e6f2465d071857854390453805#diff-a636a87d8843eeccca90140be91d4fafL43, then we can not simplify boolean condition before filter push down through join, which leads to `CartesianProduct`.

i think this change is reasonable.

**Comment by marmbrus** [comment]
Would you get the same result (and possibly more) by collapsing `operator reordering` and `constant folding` into a single batch?  This seems like a very one-off fix to me.

**Comment by DoingDone9** [comment]
yes, it can work. but this batch will have two different types of optimizers. @marmbrus

**Comment by marmbrus** [comment]
That is not fundamentally a problem.  Honestly some more thought probably needs to be put into the batches.  Really the only reasons for splitting are the following:
 - Large batches are inherently more costly as you must go through every rule, even if only a small number are making changes.  So if rules will never interact they can be in separate batches.
 - However, large batches are more powerful as there is more opportunity for rules to interact
 - Its possible for rules to undo the result of other rules.  In this case they *must* be in separate batches or it will go back and forth till the limit is reached
 - Another reason for batches is satisfying preconditions. (i.e. a plan must be analyzed before optimizing it).

**Comment by DoingDone9** [comment]
@marmbrus  /cc

**Comment by marmbrus** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -36,21 +36,20 @@ object DefaultOptimizer extends Optimizer {
     // SubQueries are only needed for analysis and can be removed before execution.
     Batch("Remove SubQueries", FixedPoint(100),
       EliminateSubQueries) ::
-    Batch("Operator Reordering", FixedPoint(100),
+    Batch("Operator Optimizations", FixedPoint(100),
```
Comment (scwf) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: how about "Operator Reordering and ConstantFolding"

**Comment by DoingDone9** [comment]
@marmbrus

**Comment by DoingDone9** [comment]
@marmbrus  @yhuai  @scwf  /cc

**Comment by marmbrus** [comment]
ok to test

**Comment by marmbrus** [comment]
Thanks!  Merged to master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -38,21 +38,20 @@ object DefaultOptimizer extends Optimizer {
...
       ProjectCollapsing,
-      CombineLimits) ::
-    Batch("ConstantFolding", FixedPoint(100),
+      CombineLimits,
       NullPropagation,
       OptimizeIn,
       ConstantFolding,
       LikeSimplification,
       BooleanSimplification,
+      PushPredicateThroughJoin,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I think we don't need to change the rule order inside a batch with fixed point. Rules within a single batch shouldn't be sensitive to execution order.

**Pull request closed without merging**